usr/lib/pkcs11/common/mech_ec.c

```diff
@@ -675,3 +691,426 @@ ec_hash_verify_final( STDLL_TokData_t      * tokdata,
+    for (i = 0; i < key_len / kdf_digest_len; i++) {
+        memcpy(ctx + z_len, &counter, sizeof(int));
+        rc = ckm_kdf(tokdata, sess, kdf, ctx, ctx_len, hash, &h_len);
+        if (rc != 0) {
+            return rc;
+        }
+        memcpy(key + i * kdf_digest_len, hash, kdf_digest_len);
+        counter++;
+    }
+
+    return CKR_OK;
+}
+
+CK_RV
+ckm_ecdh_pkcs_derive(STDLL_TokData_t *tokdata, CK_VOID_PTR other_pubkey,
+        CK_ULONG other_pubkey_len, CK_OBJECT_HANDLE base_key,
+        CK_BYTE *secret_value, CK_ULONG *secret_value_len)
+{
+    CK_RV rc;
+    CK_ATTRIBUTE *attr;
+    OBJECT *base_key_obj = NULL;
+    CK_BYTE *oid_p;
+    CK_ULONG oid_len;
+
+    if (token_specific.t_ecdh_pkcs_derive == NULL) {
+        TRACE_ERROR("ecdh pkcs derive is not supported by this token.\n");
+        return CKR_FUNCTION_NOT_SUPPORTED;
+    }
+
+    /* Find base_key struct */
+    rc = object_mgr_find_in_map1(tokdata, base_key, &base_key_obj);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Failed to acquire key from specified handle");
+        if (rc == CKR_OBJECT_HANDLE_INVALID)
+            return CKR_KEY_HANDLE_INVALID;
+        else
+            return rc;
+    }
+
+    /* Get curve oid from CKA_ECDSA_PARAMS */
+    if (!template_attribute_find(base_key_obj->template, CKA_ECDSA_PARAMS, &attr)) {
+        TRACE_ERROR("%s\n", ock_err(ERR_TEMPLATE_INCOMPLETE));
+        return CKR_TEMPLATE_INCOMPLETE;
+    }
+    oid_p = attr->pValue;
+    oid_len = attr->ulValueLen;
+
+    /* Extract EC private key (D) from base_key */
+    if (!template_attribute_find(base_key_obj->template, CKA_VALUE, &attr)) {
+        TRACE_ERROR("Could not find CKA_VALUE in the template\n");
+        return CKR_FUNCTION_FAILED;
+    }
+
+    /* Call token specific ECDH key derivation function */
+    rc = token_specific.t_ecdh_pkcs_derive(tokdata,
+            (CK_BYTE *) (attr->pValue), attr->ulValueLen,
+            (CK_BYTE *) other_pubkey, other_pubkey_len, secret_value,
+            secret_value_len, oid_p, oid_len);
+    if (rc != CKR_OK) {
+        TRACE_ERROR("Token specific ecdh pkcs derive failed with rc=%ld.\n", rc);
+        return rc;
+    }
+
+    return CKR_OK;
+}
+
+static CK_RV
+digest_from_kdf(CK_EC_KDF_TYPE kdf, CK_MECHANISM_TYPE *mech)
+{
+    switch (kdf) {
+    case CKD_SHA1_KDF:
+        *mech = CKM_SHA_1;
+        break;
+    case CKD_SHA224_KDF:
+        *mech = CKM_SHA224;
+        break;
+    case CKD_SHA256_KDF:
+        *mech = CKM_SHA256;
+        break;
+    case CKD_SHA384_KDF:
+        *mech = CKM_SHA384;
+        break;
+    case CKD_SHA512_KDF:
+        *mech = CKM_SHA512;
+        break;
+    default:
+        TRACE_ERROR("Error unsupported KDF %ld.\n", kdf);
+        return CKR_FUNCTION_FAILED;
+    }
+
+    return CKR_OK;
+}
+
+CK_RV
+pkcs_get_keytype(CK_ATTRIBUTE *attrs, CK_ULONG attrs_len,
+        CK_MECHANISM_PTR mech, CK_ULONG *type, CK_ULONG *class)
+{
+    int i;
+    CK_RV rc = CKR_TEMPLATE_INCONSISTENT;
```

Comment:
This CKR_TEMPLATE_INCONSISTENT doesn't make sense to me.
Can't we just do:
CK_RV rc = CKR_OK; 
and then remove line 946?